File: LeetCode/Medium/Maximum Length Of Concatenated String With Unique Characters/Go/maximumLengthOfConcatenatedStringWithUniqueCharacters.go
package main

import (
	"fmt"
	"strings"
)

func maxLength(arr []string) int {
	maxLength := 0
	backtrack(arr, &maxLength, 0, []string{})
	return maxLength
}

func backtrack(arr []string, maxLength *int, index int, set []string) {
	l := 0
	for i := index; i < len(arr); i++ {
		set = append(set, arr[i])
		if isUnique(strings.Join(set, "")) {
			l = len(strings.Join(set, ""))
			backtrack(arr, maxLength, i+1, set)
		}
		*maxLength = max(*maxLength, l)
		set = set[:len(set)-1]
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func isUnique(s string) bool {
	size := len(s)
	hashMap := make(map[byte]int)
	for i := 0; i < size; i++ {
		hashMap[s[i]]++
		if hashMap[s[i]] > 1 {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println(maxLength([]string{"un", "iq", "ue"}))
}

File: LeetCode/Medium/Maximum Length Of Concatenated String With Unique Characters/Java/maximumLengthOfConcatenatedStringWithUniqueCharacters.java
class Solution {
    int res = 0;

    public int maxLength(List<String> arr) {
        backtrack(arr, 0, new ArrayList<>());
        return res;    
    }

    public void backtrack(List<String> arr, int index, List<String> set) {
        int currLength = 0;
        for (int i=index; i<arr.size(); i++) {
            set.add(arr.get(i));
            String joined = String.join("", set);
            if (isUnique(joined)){
                currLength = joined.length();
                backtrack(arr, i+1, set);
            }
            res = Math.max(res, currLength);
            set.remove(set.size()-1);
        }
    }

    public boolean isUnique(String s) {
        HashMap<Character, Integer> hashMap = new HashMap<>();
        for (int i=0; i<s.length(); i++) {
            hashMap.put(s.charAt(i), hashMap.getOrDefault(s.charAt(i), 0)+1);
            if (hashMap.get(s.charAt(i)) > 1)
                return false;
        }
        return true;
    }
}
